### PR TITLE
Slackbot-like ping-on-text functionality

### PR DESCRIPTION
**Pull request by MahdiBM**
This PR adds the Slackbot's ping-on-text functionality to penny.

Members with technical roles can use ephemeral (basically private) Slash Commands, and add some texts to be pinged for.

Penny will constantly check for messages which include those texts, and DM users if needed.

**NOTEs:** 
* Need to include the `PINGS_API_BASE_URL` env variable for `PennyBOT`.
* Need to include the `BUCKET` env variable for `AutoPingLambda`.

### CODE REVIEW

**Comment by MahdiBM** [comment]
Still need to add the S3 get/post functions

**Comment by 0xTim** [comment]
I'll review once #26 is merged to avoid a duplicate review. We should add the required env vars to the stack if we need them

**Comment by MahdiBM** [comment]
What should `PINGS_API_BASE_URL` be set to then? i assume you need to create something on AWS first?

**Comment by MahdiBM** [comment]
Since i copy-pasted the pings lambda .yml file, the are some env vars left in it, including `AWS_S3_BUCKET: "penny-discord-bot"` ... can we just use that in the app instead of `BUCKET`?

**Comment by MahdiBM** [comment]
Just putting it out here that we _could_ restrict access to commands, for any reason.
For example if you think it'll be too unnecessarily heavy for Penny to to provide this functionality to everyone, we can restrict it to members with any significant roles (like backer+contributor+maintainer+moderator+core)

File: .github/workflows/pingsLambdaUpload.yml
Comment (MahdiBM) [inline .github/workflows/pingsLambdaUpload.yml:40]: mostly just copy-pasted the other lambda .yml file. please fully check this file to see if there is something that needs to be different

File: CODE/Package.swift
Comment (MahdiBM) [inline CODE/Package.swift:23]: i'll probably tag the new changes before push to main. or maybe pinning to a commit.

File: CODE/Sources/PennyBOT/+Logger.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/+Logger.swift:22]: nothing special in this file, just updating the function to accept optionals.

File: CODE/Sources/PennyBOT/BotStateManager.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/BotStateManager.swift:24]: nothing special in this file too, just updating with the new `DiscordService`

File: CODE/Sources/PennyBOT/Handlers/+String.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/+String.swift:3]: just manipulates the string to a form more appropriate for auto-ping. Makes it lowercased and removes leading/trailing spaces

File: CODE/Sources/PennyBOT/Handlers/EventHandler.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/EventHandler.swift:23]: updating the file considering the new `DiscordService`

File: CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift:10]: updating these lines of this file, considering the new `DiscordService`

File: CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift:16]: The rest of the file id basically just adding the slash commands for auto-pings, as well as refining the `link` command.

File: CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/InteractionHandler.swift:168]: Both `link` and `auto-pings` use `ephemeral` interaction messages. Meaning that the interaction-message is not visible to other people at all. Only the user of the command sees the messages.

for auto-pings this is important to make sure people don't abuse the bot by seeing what one person is registering as a word to get pinged for, then intentionally spamming that word just to ping the person.

File: CODE/Sources/PennyBOT/Handlers/MessageHandler.swift
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Handlers/MessageHandler.swift:89]: the auto-pings-checker function

File: CODE/Sources/PennySHARED/Models/S3AutoPingItems.swift
Comment (MahdiBM) [inline CODE/Sources/PennySHARED/Models/S3AutoPingItems.swift:3]: maybe we want to add something like REGEX in the future?! this enums makes it nice and possible

File: CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift
Comment (MahdiBM) [inline CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift:44]: am i doing it right?!

File: CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift
Comment (MahdiBM) [inline CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift:58]: am i doing it right?! is the `acl` right?! is the `key` something that could work?!

File: CODE/Sources/PennySHARED/Repositories/UserRepository.swift
Comment (MahdiBM) [inline CODE/Sources/PennySHARED/Repositories/UserRepository.swift:5]: just trying to keep it consistent

File: CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift
Comment (MahdiBM) [inline CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift:19]: This type was actually not needed for testing at all, so I removed it.

File: CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift
Comment (MahdiBM) [inline CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift:47]: add `auto-pings` command to the register-test

File: CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift
Comment (MahdiBM) [inline CODE/Tests/PennyBOTTests/GatewayProcessingTests.swift:212]: sends an event, then checks to make sure DM messages have been sent to people who had registered to get pinged for words

File: CODE/task-def.json
Comment (MahdiBM) [inline CODE/task-def.json:32]: Need to add auto-pings lambda URL

**Review by MahdiBM — COMMENTED** [review]
I've figured self-reviewing is quite helpful for the other reviewers.
only 27 comments 😝

**Comment by MahdiBM** [comment]
now this PR also resolves #38

**Comment by MahdiBM** [comment]
Since i did this after the comments, i couldn't have mentioned it, but i've removed passing of `Logger` in a lot of places in PennyBOT where it wasn't actually needed. Since we bootstrap the logging system with the Logger we want, we can just initialize a new logger instead of requiring it as a parameter in places like `EventHandler(...., logger: logger)`.

File: CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift
Comment (0xTim) [reply CODE/Sources/PennySHARED/Repositories/S3AutoPingsRepository.swift:58]: Yeah ACL is right since the only thing reading and writing should be the lambda.
The key is essentially the path in the bucket

**Review by 0xTim — APPROVED** [review]
This looks good! Happy to get it deployed and start testing it!

**Pull request merged**
